Setup command throws error messages for repos that already have commits

Scenario - student repos have been created via setup. Students work on repos. Setup command is run again (for latecomers typically), giving this message for some repos that have now had commits made by students:

```
...
[ERROR] Failed to push to https://gits-15.sys.kth.se/inda-20/XXXX-task-1
return code: 1
To https://gits-15.sys.kth.se/inda-20/XXXX-task-1
[WARNING] 13 pushes failed ...
```

Perhaps a nicer message?

--- a/src/_repobee/command/repos.py
+++ b/src/_repobee/command/repos.py
@@ -230,14 +230,9 @@ def update_student_repos(
             template_repos, api, cwd=pathlib.Path(tmpdir)
         )
 
-        # we want to exhaust this iterator immediately to not have progress
-        # bars overlap
-        fetched_teams = list(progresswrappers.get_teams(teams, api))
-
-        push_tuple_iter = _create_push_tuples(
-            fetched_teams, template_repos, api
+        push_tuple_iter = _create_update_push_tuples(
+            teams, template_repos, api
         )
-
         push_tuple_iter_progress = plug.cli.io.progress_bar(
             push_tuple_iter,
             desc="Setting up student repos",
@@ -254,6 +249,35 @@ def update_student_repos(
 
     plug.log.info("Done!")
     return hook_results
+
+
+def _create_update_push_tuples(
+    teams: Iterable[plug.StudentTeam],
+    template_repos: Iterable[plug.TemplateRepo],
+    api: plug.PlatformAPI,
+) -> Iterable[Push]:
+    """Create push tuples for existing repos. Repos that don't exist are
+    ignored.
+
+    Args:
+        teams: An iterable of teams.
+        template_repos: Template repositories.
+        api: A platform API instance.
+    Returns:
+        A list of Push namedtuples for all student repo urls that relate to
+        any of the master repo urls.
+    """
+    urls_to_templates = {}
+    for team, template_repo in itertools.product(teams, template_repos):
+        repo_url, *_ = api.get_repo_urls(
+            [template_repo.name], team_names=[team.name]
+        )
+        urls_to_templates[repo_url] = template_repo
+
+    for repo in api.get_repos(urls_to_templates.keys()):
+        template = urls_to_templates[repo.url]
+        branch = git.active_branch(template.path)
+        yield git.Push(template.path, repo.url, branch)
 
 
 def _open_issue_by_urls(

--- a/src/_repobee/command/repos.py
+++ b/src/_repobee/command/repos.py
@@ -18,7 +18,7 @@ import re
 import os
 import sys
 import tempfile
-from typing import Iterable, List, Optional, Mapping, Union
+from typing import Iterable, List, Optional, Mapping, Union, Tuple
 
 import repobee_plug as plug
 
@@ -108,11 +108,12 @@ def _create_push_tuples(
     template_repos: Iterable[plug.TemplateRepo],
     api: plug.PlatformAPI,
 ) -> Iterable[Push]:
-    """
+    """Create push tuples for newly created repos. Repos that already exist are
+    ignored.
+
     Args:
         teams: An iterable of teams.
-        authed_master_urls: An iterable of authenticated master repo urls.
-        master_repo_paths: A mapping (master_repo_url -> master_repo_path).
+        template_repos: Template repositories.
         api: A platform API instance.
     Returns:
         A list of Push namedtuples for all student repo urls that relate to
@@ -120,7 +121,7 @@ def _create_push_tuples(
     """
     for team, template_repo in itertools.product(teams, template_repos):
         repo_name = plug.generate_repo_name(team, template_repo.name)
-        repo = _create_or_fetch_repo(
+        created, repo = _create_or_fetch_repo(
             name=repo_name,
             description=f"{repo_name} created for {team.name}",
             private=True,
@@ -128,11 +129,12 @@ def _create_push_tuples(
             api=api,
         )
 
-        yield git.Push(
-            local_path=template_repo.path,
-            repo_url=api.insert_auth(repo.url),
-            branch=git.active_branch(template_repo.path),
-        )
+        if created:
+            yield git.Push(
+                local_path=template_repo.path,
+                repo_url=api.insert_auth(repo.url),
+                branch=git.active_branch(template_repo.path),
+            )
 
 
 def _create_or_fetch_repo(
@@ -141,10 +143,13 @@ def _create_or_fetch_repo(
     private: bool,
     api: plug.PlatformAPI,
     team: Optional[plug.Team] = None,
-) -> plug.Repo:
+) -> Tuple[bool, plug.Repo]:
     try:
-        return api.create_repo(
-            name, description=description, private=private, team=team
+        return (
+            True,
+            api.create_repo(
+                name, description=description, private=private, team=team
+            ),
         )
     except plug.PlatformError:
         team_name = team.name if team else None
@@ -154,7 +159,7 @@ def _create_or_fetch_repo(
             api.assign_repo(
                 team=team, repo=repo, permission=plug.TeamPermission.PUSH
             )
-        return repo
+        return False, repo
 
 
 def _clone_all(
@@ -351,7 +356,7 @@ def migrate_repos(
             plug.TemplateRepo(
                 name=repo.name, url=repo.url, _path=workdir / repo.name
             )
-            for repo in create_repo_it
+            for _, repo in create_repo_it
         ]
 
         git.push(

--- a/src/_repobee/command/repos.py
+++ b/src/_repobee/command/repos.py
@@ -277,7 +277,7 @@ def _create_update_push_tuples(
     for repo in api.get_repos(urls_to_templates.keys()):
         template = urls_to_templates[repo.url]
         branch = git.active_branch(template.path)
-        yield git.Push(template.path, repo.url, branch)
+        yield git.Push(template.path, api.insert_auth(repo.url), branch)
 
 
 def _open_issue_by_urls(

--- a/src/repobee_testhelpers/funcs.py
+++ b/src/repobee_testhelpers/funcs.py
@@ -14,11 +14,21 @@ from repobee_testhelpers import localapi
 from repobee_testhelpers import const
 
 
-def initialize_repo(path: pathlib.Path) -> git.Repo:
-    """Initialize the directory to a Git repo and commit all files in it."""
+def initialize_repo(
+    path: pathlib.Path, default_branch: str = "master"
+) -> git.Repo:
+    """Initialize the directory to a Git repo and commit all files in it.
+
+    Args:
+        path: Path to the directory to turn into a Git repository.
+        default_branch: Name of the default branch to use.
+    Returns:
+        A repo.
+    """
     repo = git.Repo.init(path)
     repo.git.config("user.name", const.TEACHER)
     repo.git.config("user.email", f"{const.TEACHER}@repobee.org")
+    repo.git.checkout("-b", default_branch)
     repo.git.add(".", "--force")
     repo.git.commit("-m", "Initial commit")
     return repo

--- a/tests/new_integration_tests/test_repos.py
+++ b/tests/new_integration_tests/test_repos.py
@@ -161,6 +161,35 @@ class TestSetup:
                 repo_dict,
             )
 
+    def test_does_not_push_to_existing_repos(
+        self, platform_url, with_student_repos, capsys
+    ):
+        """This command should not push to existing repos, that's for the
+        ``update`` command to do.
+        """
+        with tempfile.TemporaryDirectory() as tmpdir:
+            # arrange
+            template_repo_dir = pathlib.Path(tmpdir)
+            task = template_repo_dir / TEMPLATE_REPO_NAMES[0]
+            create_local_repo(task, [("best/file/ever.txt", "content")])
+
+            # act
+            # this push would fail if it was attempted, as the repo
+            # content of the local template does not match that of
+            # the remote template
+            funcs.run_repobee(
+                f"repos setup -a {TEMPLATE_REPOS_ARG} "
+                f"--base-url {platform_url} "
+                "--allow-local-templates",
+                workdir=template_repo_dir,
+            )
+
+        # nothing should have changed, and there should be no errors
+        assert_student_repos_match_templates(
+            STUDENT_TEAMS, TEMPLATE_REPO_NAMES, funcs.get_repos(platform_url),
+        )
+        assert "[ERROR]" not in capsys.readouterr().out
+
     def test_setup_with_local_repos_fails_without_local_templates_arg(
         self, platform_url
     ):
@@ -285,6 +314,25 @@ class TestClone:
         assert not expected_repo_names
 
 
+class TestUpdate:
+    """Tests for the ``repos update`` command."""
+
+    def test_does_not_create_repos(self, platform_url):
+        """This command should only update existing repos, it's not allowed to
+        create repos.
+        """
+        # arrange, must create the student teams
+        funcs.run_repobee(f"teams create --base-url {platform_url}")
+
+        # act
+        funcs.run_repobee(
+            f"repos update -a {TEMPLATE_REPOS_ARG} --base-url {platform_url}"
+        )
+
+        # assert
+        assert not funcs.get_repos(platform_url)
+
+
 class TestMigrate:
     """Tests for the ``repos migrate`` command."""
 
@@ -315,14 +363,6 @@ class TestMigrate:
             assert repo.branches[0].name == strange_branch_name
 
 
-def reponize(dirpath: pathlib.Path, default_branch: str):
-    """Turn a non-empty directory into a Git repo and commit all content."""
-    repo = git.Repo.init(dirpath)
-    repo.git.checkout("-b", default_branch)
-    repo.git.add(".", "--force")
-    repo.git.commit("-m", "Initial commit")
-
-
 def create_local_repo(
     path: pathlib.Path,
     files: Iterable[Tuple[str, str]],
@@ -330,12 +370,22 @@ def create_local_repo(
 ) -> str:
     """Create a local repository in the provided basedir and return a
     hash of the contents.
+
+    Args:
+        path: Path to put the repository at. Parent directories are created if
+            they don't exist.
+        files: Files to add to the repository. Should be tuples on the form
+            (relpath, content), where relpath is a filepath relative to the
+            root of the repository.
+        default_branch: The default branch to use for the repository.
+    Returns:
+        The sha1 hash of the repository.
     """
-    path.mkdir()
+    path.mkdir(parents=True)
     for filename, content in files:
         file = path / filename
         file.parent.mkdir(parents=True, exist_ok=True)
         (path / filename).write_text(content, encoding="utf8")
     sha = funcs.hash_directory(path)
-    reponize(path, default_branch)
+    funcs.initialize_repo(path, default_branch)
     return sha

